nominal/nominal.py

```diff
@@ -1,628 +1,268 @@
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def upload_polars(
+    df: pl.DataFrame,
+    name: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    description: str | None = None,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a polars.DataFrame."""
+    conn = get_default_connection()
+
+    def write_and_close(df: pl.DataFrame, w: BinaryIO) -> None:
+        df.write_csv(w)
+        w.close()
+
+    with reader_writer() as (reader, writer):
+        # write the dataframe to CSV in another thread
+        t = Thread(target=write_and_close, args=(df, writer))
+        t.start()
+        dataset = conn.create_dataset_from_io(
+            reader,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=FileTypes.CSV,
+            description=description,
+        )
+        t.join()
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def upload_csv(
+    path: Path | str,
+    name: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    description: str | None = None,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a .csv or .csv.gz file"""
+    path = Path(path)
+    file_type = FileType.from_path_dataset(path)
+    conn = get_default_connection()
+    with open(path, "rb") as f:
+        dataset = conn.create_dataset_from_io(
+            f,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=file_type,
+            description=description,
         )
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def get_dataset(rid: str) -> Dataset:
+    """Retrieve a dataset from the Nominal platform by its RID."""
+    conn = get_default_connection()
+    return conn.get_dataset(rid)
+
+
+def create_run(
+    title: str,
+    start: datetime | str | IntegralNanosecondsUTC,
+    end: datetime | str | IntegralNanosecondsUTC,
+    description: str | None = None,
+) -> Run:
+    conn = get_default_connection()
+    return conn.create_run(
+        title,
+        "" if description is None else description,
+        start=_parse_timestamp(start),
+        end=_parse_timestamp(end),
+    )
+
+
+def get_run(rid: str) -> Run:
+    """Retrieve a run from the Nominal platform by its RID."""
+    conn = get_default_connection()
+    return conn.get_run(rid)
+
+
+def search_runs(
+    *,
+    start: str | datetime | IntegralNanosecondsUTC | None = None,
+    end: str | datetime | IntegralNanosecondsUTC | None = None,
+    exact_title: str | None = None,
+    label: str | None = None,
+    property: tuple[str, str] | None = None,
+) -> list[Run]:
```

Comment:
what will running `search_runs` with no parameters return? good to avoid the case where users call the method accidentally and it suddenly blocks for ages trying to read every single run

Reply:
it actually throws, conjure doesn't allow it

Reply:
(but this is also why I prefer the `-> Iterable` signatures!)

Reply:
I now throw if none are provided